net: Log to net category for exceptions in ProcessMessages

laanwj opened the pull request:
Remove the forest of special exceptions based on string matching, and simply log a short message to the NET logging category when an exception happens during packet processing. It is not good to panick end users with verbose errors (let alone writing to stderr) when any peer can generate them.

laanwj committed:
net: Log to net category for exceptions in ProcessMessages

Remove the forest of special exceptions, and simply log a short
message to the NET logging category when an exception happens during
packet processing. It is not good to panick end users with errors
that any peer can generate (let alone writing to stderr).

practicalswift commented:
Tested ACK 4d88c3dcb61e7c075ed3dd442044e0eff4e3c8de

MarcoFalke commented:
Concept ACK. I probably should have done this in #16021.

promag commented on src/net_processing.cpp:
```diff
@@ -3333,32 +3333,10 @@ bool PeerLogicValidation::ProcessMessages(CNode* pfrom, std::atomic<bool>& inter
+        LogPrint(BCLog::NET, "%s(%s, %u bytes): Exception '%s' (%s) caught\n", __func__, SanitizeString(strCommand), nMessageSize, e.what(), typeid(e).name());
     } catch (...) {
-        PrintExceptionContinue(nullptr, "ProcessMessages()");
+        LogPrint(BCLog::NET, "%s(%s, %u bytes): Unknown exception caught\n", __func__, SanitizeString(strCommand), nMessageSize);
```
nit, `unexpected` or `unhandled` instead or unknown?

laanwj replied:
It's really *unknown*. All known exceptions go to the other path. If it ends up here, some really weird code threw an integer or bare string or such.

promag reviewed (commented):
ACK 4d88c3dcb61e7c075ed3dd442044e0eff4e3c8de.

jnewbery reviewed (commented):
Concept ACK. One question about `typeid()` inline.

MarcoFalke commented:
ACK 4d88c3dcb61e7c075ed3dd442044e0eff4e3c8de 👃

Tested both cases:

* Normal exception:

 - 2019-12-20T16:50:10.551972Z [msghand] received: tx (85 bytes) peer=2
 - 2019-12-20T16:50:10.552100Z [msghand] ProcessMessages(tx, 85 bytes): Exception 'Superfluous witness record: iostream error' (NSt8ios_base7failureB5cxx11E) caught
 - 2019-12-20T16:50:10.552120Z [msghand] ProcessMessages(tx, 85 bytes) FAILED peer=2

* Something else:

 - 2019-12-20T17:03:18.898307Z [msghand] received: tx (85 bytes) peer=2
 - 2019-12-20T17:03:18.898406Z [msghand] ProcessMessages(tx, 85 bytes): Unknown exception caught
 - 2019-12-20T17:03:18.898421Z [msghand] ProcessMessages(tx, 85 bytes) FAILED peer=2

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK 4d88c3dcb61e7c075ed3dd442044e0eff4e3c8de 👃

Tested both cases:

* Normal exception:

 - 2019-12-20T16:50:10.551972Z [msghand] received: tx (85 bytes) peer=2
 - 2019-12-20T16:50:10.552100Z [msghand] ProcessMessages(tx, 85 bytes): Exception 'Superfluous witness record: iostream error' (NSt8ios_base7failureB5cxx11E) caught
 - 2019-12-20T16:50:10.552120Z [msghand] ProcessMessages(tx, 85 bytes) FAILED peer=2

* Something else:

 - 2019-12-20T17:03:18.898307Z [msghand] received: tx (85 bytes) peer=2
 - 2019-12-20T17:03:18.898406Z [msghand] ProcessMessages(tx, 85 bytes): Unknown exception caught
 - 2019-12-20T17:03:18.898421Z [msghand] ProcessMessages(tx, 85 bytes) FAILED peer=2
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUjeMwwArad7Z2G99LooWjtt7/fr7X2kNavsjZh1YvI8AytS/9uZkURdVFg96geY
Q3jJqisfbeNTwJm0zsNbSWChN5OlWrZ8naGMDfyALM0oQXzj5rgCZM5WIZta4Me9
XDFq4upRfvg5QhmFI1feGX4mIYqdf2YlxYjkLz293rvTRTnQsEJ655tk1uWrm1jK
SnGKtv29IKNO0yEb4WOE4pB7muzlf3mXVO9K+21HglK5qY3Opy7hkO8NsJSpilvz
Gr4Y8yJFhJMWL8gG3fZb/YaViDzcNanbjMNXTUHIwMQZ6sQhOpDvXKYNvyNJG8Di
1xleSq1lEHp/nAdPor29fMCLVaSi7EY7A32LetW7SE30I6zbqlbH3XW4rg/mgiHr
bzoz/yJT6vrxRm8BOLYu9PjQLhKOVeDL0iGmxQaHxUdnpDdlQUYgjTeCcGUFMptB
vwD7xTEfgFdeBQuDjwFccpDDgUdUO8FBtFBHxxOEa1VVHtEvteRRvMzXCnJFaR0K
NONhQti2
=lSTE
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `df6e3a596aced982ab075caecb0b392dea1ae5573fef22c74d4cdd198ffc188c  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108df6e3a596aced982ab075caecb0b392dea1ae5573fef22c74d4cdd198ffc188cf010d45d66122b30c694a61de7b9c222d2be08fff0104f675254a9a839c377b16cee1e17593108f1045dfcff18f008e2424083dadf3c740083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff010609d28427843b979f1e3195cf414e6ce08f1045dfcff18f00837f259f4d6fbccc70083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff010c47ee0a6e8a32ccec3980c9d694c21fb08f1045dfcff19f0082887de7986f5a06a0083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6df010f123bfdbfbfe33d370aebe2cffc74c1f08f1045dfcff19f0089a7fe4b2ae0725680083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>

hebasto commented:
Concept ACK.

hebasto commented on src/net_processing.cpp:
```diff
@@ -3333,32 +3333,10 @@ bool PeerLogicValidation::ProcessMessages(CNode* pfrom, std::atomic<bool>& inter
-    catch (const std::exception& e) {
-        PrintExceptionContinue(&e, "ProcessMessages()");
+    } catch (const std::exception& e) {
+        LogPrint(BCLog::NET, "%s(%s, %u bytes): Exception '%s' (%s) caught\n", __func__, SanitizeString(strCommand), nMessageSize, e.what(), typeid(e).name());
```
And [`name()`](https://en.cppreference.com/w/cpp/types/type_info/name):
> Returns an _implementation defined_ null-terminated character string containing the name of the type. _No guarantees are given_; in particular, the returned string can be identical for several types and change between invocations of the same program.

On Linux Mint 19.3 (GCC 7.4.0) with included `<typeinfo>` the output remains the same: something like `...(NSt8ios_base7failureB5cxx11E)...`.

hebasto replied:
Also we use the same approach here (line 544):
https://github.com/bitcoin/bitcoin/blob/0cda5573405d75d695aba417e8f22f1301ded001/src/util/system.cpp#L542-L547

laanwj replied:
This was simply borrowed from the exception logging function that we've already used since Satoshi's days. I think the idea is that it shows the type in case the exception message is not informative, I'm ok with removing it, no opinion, but I don't think it hurts…

laanwj replied:
I'll add a commit that adds `#include <typeinfo>` to here and system.cpp. I think that's a good point, though no C++ compiler seems to have stumbled over it.

laanwj committed:
Add missing typeinfo includes

The use of `typeid()` for logging exception types requires this include
according to https://en.cppreference.com/w/cpp/language/typeid.

MarcoFalke commented:
re-ACK 4bdd68f301a9cee3360deafc7531c638e923226b (only change is adding includes) 🕕

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

re-ACK 4bdd68f301a9cee3360deafc7531c638e923226b (only change is adding includes) 🕕
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUguNgwAoLtUwlt9Rl8Fd4Kkhw+iyyVzIoIfk+HL6LcG9BqfDfttfcijBiynEtNV
29HzFgff3tiBWuYeN013IqeW1rE3MYxI4pDpGqeaWKkoj9YA6PW/jPWj8okAGzAC
aYsI/2cT/H6Ea/WL7foSeMejRryRLQfD8Jg241g4Rm8iPEm0U0Tqgk5s9oqzDfMi
J1R8Et1yDZo6+enk03ytF1WW3X+SN5PkI6M4CfShnf4ZxRyQ8atYUhRMOfOb8xFA
vu2HKtYE+oWmFTfR1j6k5A9VO5ezOeL24nF8XP8FV+bcH7evzMOvudwdNMVuKag/
giW1i/ElaqHLZvTXxV1jGcu4/Vu1NfvylHQLuOft8H2BORe+OBrXAXT+qIpG5DtA
q2xyhHCfkJwItwQAMGYhauHISq93NA77CNENxyxKPvupAMfwuViYYYG+gQ8cOXg9
5YQ7GhkTFtCrTqU7abF/66XU62NKAEXvzJCjlF+/WzILoPncYAjTQcNO71pHMKYL
lM2l3l5j
=8t8Y
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `f5a70ac73ac8d01fd424479a9ace0934a6de8392add537503e92e054731e2b32  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108f5a70ac73ac8d01fd424479a9ace0934a6de8392add537503e92e054731e2b32f010acc9465022a02dc32aad2d5337bbcc2708fff010833031a0363ff77fb99a34f5a3019bc308f1045e0e4f40f0089236b85d2385fe5c0083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6df010e5359b1732cb7e5f8b4be86661f790f508f1045e0e4f40f008883a9c0cf3e600b70083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>

promag commented:
ACK 4bdd68f301a9cee3360deafc7531c638e923226b, could squash.

fanquake commented:
Being backported in 17858.

practicalswift commented:
Reviewers of this PR might want to review the somewhat related PR #17828 ("net: Use log categories when logging events that P2P peers can trigger arbitrarily").

fanquake commented:
This was backported to the 0.18 branch in #17974.
